src/jenkins.py

```diff
@@ -705,3 +709,195 @@ def get_agent_name(unit_name: str) -> str:
+            dependencies.
+        dependency_lookup: The plugin dependency lookup table.
+
+    Yields:
+        The allowed plugin short name.
+    """
+    seen: set[str] = set()
+    for plugin in top_level_plugins:
+        if plugin not in seen:
+            yield from _traverse_dependencies(plugin, dependency_lookup, seen)
+
+
+def _get_top_level_plugins(
+    plugins: typing.Iterable[str], dependency_lookup: typing.Mapping[str, typing.Iterable[str]]
+) -> typing.Iterable[str]:
+    """Get top level plugins that are not dependencies of other plugins.
+
+    Args:
+        plugins: Plugins to extract top level plugins from.
+        dependency_lookup: The dependency lookup table.
+
+    Returns:
+        All plugins that are not dependency of another plugin.
+    """
+    dependent_plugins: set[str] = set()
+    for _, dependencies in dependency_lookup.items():
+        dependent_plugins = dependent_plugins.union(dependencies)
+
+    return set(plugins) - dependent_plugins
+
+
+def remove_unlisted_plugins(
+    plugins: typing.Iterable[str] | None,
+    container: ops.Container,
+    client: jenkinsapi.jenkins.Jenkins | None = None,
+) -> None:
+    """Remove plugins that are not in the list of desired plugins.
+
+    Args:
+        plugins: The list of plugins that can be installed.
+        container: The workload container.
+        client: The Jenkins API client.
+
+    Raises:
+        JenkinsPluginError: if there was an error removing unlisted plugin.
+        JenkinsError: if there was an error restarting Jenkins after removing the plugin.
+        TimeoutError: if it took too long to restart Jenkins after removing the plugin.
+    """
+    if not plugins:
+        return
+
+    client = client if client is not None else _get_client(get_admin_credentials(container))
+    res = client.run_groovy_script(
+        """
+def plugins = jenkins.model.Jenkins.instance.getPluginManager().getPlugins()
+plugins.each {
+    println "${it.getShortName()} (${it.getVersion()}) => ${it.getDependencies()}"
+}
+"""
+    )
+    dependency_lookup = _build_dependencies_lookup(res.splitlines())
+    allowed_plugins = _get_allowed_plugins(
+        itertools.chain(plugins, REQUIRED_PLUGINS), dependency_lookup
+    )
+    plugins_to_remove = set(dependency_lookup.keys()) - set(allowed_plugins)
+    if not plugins_to_remove:
+        return
+
+    try:
+        client.delete_plugins(plugin_list=plugins_to_remove, restart=False)
+    except jenkinsapi.custom_exceptions.JenkinsAPIException as exc:
+        logger.error("Failed to remove the following plugins: %s, %s", plugins_to_remove, exc)
+        raise JenkinsPluginError("Failed to remove plugins.") from exc
+    logger.debug("Removed %s", plugins_to_remove)
+
+    top_level_plugins = _get_top_level_plugins(plugins_to_remove, dependency_lookup)
+    _set_jenkins_system_message(
+        message="The following plugins have been removed by the system administrator: "
```

Comment:
Perhaps include the reason why they have been removed?

Reply:
I think the line after next "To allow the plugins, please include them in the plugins configuration of the charm." is quite explanatory. Shall I move this message up?